### ISSUE TITLE
Menu hide/show widget not working

### ISSUE THREAD

**Issue by mlcampbe** [felixhageloh/uebersicht#269]
I have several widgets setup in my widget folder. Of these I turn some on and some off. At some point later I want to either enable or disable a widget so I go to the menu and select the 'hide widget' option. No matter if it is hidden currently and I want to show it or if it is shown currently and I want to hide it nothing happens.

Oddly though if I stop/restart ubersicht completely then I can hide/show widgets successfully. Only after some time does it stop working.

Any thoughts?

**Comment by felixhageloh** [comment]
do you still see this happening?

**Comment by mlcampbe** [comment]
Felix,

Yes afraid so. It seems to me that it starts after the machine goes into sleep mode and upon waking it up I can’t enable/disable specific widgets until I do a ‘refresh all’.

— 
Mike Campbell
mlcampbell@gmail.com

**Comment by felixhageloh** [comment]
I see! If I understand correctly, widgets are showing and working fine otherwise, but you can't enable/disable them anymore?
Are there any errors in the debug console or in the Console app?

**Comment by mlcampbe** [comment]
Felix,

Correct, I have several widgets running and that all work fine and refresh as scheduled. After the machine sleeps/wakes up these widgets continue to work as expected. If I stop/restart uebersicht then I can use the system menu, select a widget, and click the hide widget option and it will work fine as well at that time. However, some time later (and I think after the next sleep/wake up) when I click the hide widget option nothing happens. The option does not get checked (or unchecked) and the widget never shows or goes away. It is like I never tried at all. If I then select the refresh all option I used to think that the hide widget option would start working but that does not look to be the case. I have to physically top/restart uebersicht completely to make it start working again.

I see nothing in the console window nor in the OS X console app.

— 
Mike Campbell
mlcampbell@gmail.com

**Comment by felixhageloh** [comment]
I have a new build here https://www.dropbox.com/s/kcjohvajbdymu6m/Uebersicht-1.2.49.app.zip with better error logging. 
Can you give it a try and check the console(s) again? Hopefully we will see an error now if there is any

**Comment by mlcampbe** [comment]
Felix,

I have installed the new build and will test it out. I will note that even with this build when I hide/show a widget from the menu it is working since I just restarted the app. However, I do not see anything printed to the ubersicht console when doing so. Are you expecting to see some output in the uberischt console even when it is working?

— 
Mike Campbell
mlcampbell@gmail.com

**Comment by mlcampbe** [comment]
Felix,

I came back from lunch and the machine had gone to sleep. I woke it up and opened the debug console and then tried to enable a widget. It did not enable and I got nothing in the console. It is almost like I am not even selecting the ‘hide widget’ option at all.

— 
Mike Campbell
mlcampbell@gmail.com

**Comment by felixhageloh** [comment]
thanks for the info!
Do you see anything in Console.app tho? (you can search for 'sicht' to only show output from Übersicht)

**Comment by mlcampbe** [comment]
Felix,

No, nothing there either. I found one message that said “using 1 screen” but that was not related to hide/unhide of a widget. I restarted ubersicht with the console.app running and I see startup messages with the last being "loaded http://127.0.0.1:41416/69678529 <http://127.0.0.1:41416/69678529>” but hide/show of a widget (which now works due to the restart) does not print anything.

— 
Mike Campbell
mlcampbell@gmail.com

**Comment by felixhageloh** [comment]
Thanks again. Still haven't made any progress on this. Can I ask which widgets you are using - maybe I can reproduce it that way.

**Comment by mlcampbe** [comment]
Felix,

Are you asking which ones I am trying to hide/show or all of the ones I am using? It does not matter which I try to hide or show they all have the same behavior. As for the list of widgets they are mostly ones I have created. Here is a screenshot of the dropdown showing what I have installed.



— 
Mike Campbell
mlcampbell@gmail.com

**Comment by felixhageloh** [comment]
yeah I did mean widgets you have installed in general - could you send your screenshot again? It doesn't show for some reason

**Comment by mlcampbe** [comment]
Felix,

Here is the screenshot again. If you can’t see it they try accessing it at https://cl.ly/0n0p1o0d0i3d <https://cl.ly/0n0p1o0d0i3d> as I have uploaded it there too.

— 
Mike Campbell
mlcampbell@gmail.com

**Comment by felixhageloh** [comment]
thanks! It is still not inlined (maybe because you are responding by mail?), but the link works!

**Comment by felixhageloh** [comment]
Hi there, coming back to this: I was in the process of installing your widgets to reproduce this, but I realized that I had misread one of your comments and that some widgets are not from the gallery. 

Could you provide me with zip of your entire widget folder? That way I can just install all your widgets in one go and hopefully reproduce this

**Comment by mlcampbe** [comment]
Felix,

Most of the widgets are not from the gallery and are my homemade ones. And, on top of that in a couple of widgets I use the “command” option to call a perl script to do some work. For example, my weather widget calls weather.pl that retrieves the weather, builds an icon, and output the text which the widget then formats.

I have included a zip of the widgets folder but not all widgets are in use. Normally I am only running the following:
bottomshade
devicebatt
ical
oracle-ip
potd
weather

Of the above potd and weather call perl scripts which I have included but edited them so that my api keys to the 3rd party services have been removed. Let me know if you have any questions.

I will send this in separate emails as when I sent them all in one it was rejected due to size limits. So this email will contain the widgets and potd.pl script and a 2nd email I will try to send the weather scripts.



— 
Mike Campbell
mlcampbell@gmail.com

**Comment by mlcampbe** [comment]
Felix,

Here are the weather scripts.



— 
Mike Campbell
mlcampbell@gmail.com

**Comment by hawkrives** [comment]
@mlcampbe FYI, Github's not passing your email attachments through into the issue itself. You might need to either email them to @felixhageloh directly or come attach them to an issue comment through the web interface.

**Comment by mlcampbe** [comment]
I've gotten back around to testing this out and moved all of my self-built widgets out of the widgets folder with the exception of 1 widget and that is the Simple Keyboard/Mouse Stats one from the http://tracesof.net/uebersicht-widgets/ site. I am still seeing the same behavior in that after the machine sleeps I see 2 problems:
1 - Using the system menu I cannot enable/disable any of the widgets that are showing up there
2 - If I edit any of the widgets and save the file the changes are not reflected in the running widget.

If I stop/restart the entire uebersicht program then both of the above are resolved until the machine sleeps and re-wakes.

Note I am currently using version 1.2 (53)

**Comment by DavideFauri** [comment]
Hello,

I can confirm that I have the same issues:
1 - I cannot enable, disable, hide or refresh any widget using the menu in the top right bar
2 - After launching Ubersicht and waiting a short while, the server at port 46146 is unresponsive, disabling the interaction key and "freezing" the widgets in place
3 - 

I've attached the result of my console output, in which I start the application at 15:06, I close the lid of my laptop at 15:08 and then open it again at 15:08:09
The relevant part, as far as I understand, starts at 15:08:41 where the socket receives a DISCONNECTED event.

[console_out_ubersicht.log](https://github.com/felixhageloh/uebersicht/files/2538814/console_out_ubersicht.log)

Moreover, from a fresh start of the application, if I try to refresh all the widgets, the console keeps repeating this message at 5 seconds intervals:

`Error loading webview: Error Domain=NSURLErrorDomain Code=-999 "(null)" UserInfo={NSErrorFailingURLStringKey=http://127.0.0.1:41416/722473485, NSErrorFailingURLKey=http://127.0.0.1:41416/722473485, _WKRecoveryAttempterErrorKey=<WKReloadFrameErrorRecoveryAttempter: 0x600001781860>}`

**Comment by DavideFauri** [comment]
So, after a bit of debugging I found out the cause of my problem:

I setup one of my widgets to create a link for a random .pdf file in my Downloads folder, so that when I'm bored I can have a suggestion of what to read from my *enormous* backlog.
To do so, I had to find a way to access files outside of the Ubersicht scripts folder.
I then thought it was a smart idea to create a symbolic link with `ln -s ~/Downloads` in the scripts folder, so that the script would traverse the path, look for my file and link to it.
What I didn't foresee, is Ubersicht pre-loading the entire contents of the Downloads folder in memory, making the server crash everytime, and making the application unresponsive.

I've removed the link and now everything works again. Sorry for the bother -_-'
